k2/csrc/fsa_algo.cc

```diff
@@ -150,6 +166,135 @@ void Connect(const Fsa &a, Fsa *b, std::vector<int32_t> *arc_map /*=nullptr*/) {
   }
 }
 
+bool Intersect(const Fsa &a, const Fsa &b, Fsa *c,
+               std::vector<int32_t> *arc_map_a /*= nullptr*/,
+               std::vector<int32_t> *arc_map_b /*= nullptr*/) {
+  CHECK_NOTNULL(c);
+  c->arc_indexes.clear();
+  c->arcs.clear();
+  if (arc_map_a != nullptr) arc_map_a->clear();
+  if (arc_map_b != nullptr) arc_map_b->clear();
+
+  if (IsEmpty(a) || IsEmpty(b)) return true;
+  if (!IsArcSorted(a) || !IsArcSorted(b)) return false;
+  // either `a` or `b` must be epsilon-free
+  if (!IsEpsilonFree(a) && !IsEpsilonFree(b)) return false;
+
+  int32_t final_state_a = a.NumStates() - 1;
+  int32_t final_state_b = b.NumStates() - 1;
+  const auto arc_a_begin = a.arcs.begin();
+  const auto arc_a_end = a.arcs.end();
+  const auto arc_b_begin = b.arcs.begin();
+  const auto arc_b_end = b.arcs.end();
+  using ArcIterator = std::vector<Arc>::const_iterator;
+
+  const int32_t final_state_c = -1;  // just as a placeholder
+  // no corresponding arc mapping from `c` to `a` or `c` to `b`
+  const int32_t arc_map_none = -1;
+  auto &arc_indexes_c = c->arc_indexes;
+  auto &arcs_c = c->arcs;
+
+  // map state pair to unique id
+  std::unordered_map<StatePair, int32_t, PairHash> state_pair_map;
+  std::queue<StatePair> qstates;
+  qstates.push({0, 0});
+  state_pair_map.insert({{0, 0}, 0});
+  state_pair_map.insert({{final_state_a, final_state_b}, final_state_c});
+  int32_t state_index_c = 0;
+  while (!qstates.empty()) {
+    arc_indexes_c.push_back(static_cast<int32_t>(arcs_c.size()));
+
+    auto curr_state_pair = qstates.front();
+    qstates.pop();
+    // as we have inserted `curr_state_pair` before.
+    int32_t curr_state_index = state_pair_map[curr_state_pair];
+
+    auto state_a = curr_state_pair.first;
+    ArcIterator a_arc_iter_begin = arc_a_begin + a.arc_indexes[state_a];
+    ArcIterator a_arc_iter_end =
+        (state_a != final_state_a) ? (arc_a_begin + a.arc_indexes[state_a + 1])
```

Comment:
I think we should make the arc_indexes array one larger, with a duplicate of the final number, to avoid these checks.  
Again, it won't require much work as long as you modify the constructor to not take the arc_indexes arg, and as long as you always call NumStates() to get the number of states, rather than using arc_indexes.size() (it will have to subtract one).

Reply:
OK, will add that state and check all of implemented algortithms.